### ISSUE TITLE
Upgrade to 0.27.1 failed for foreign key constraint "fk_report_card_ref_database_id"

### ISSUE THREAD

**Issue by sailxjx** [metabase/metabase#6520]
- Your databases: (e.x. MySQL, Postgres, MongoDB, …)
Postgres

- Metabase version: (e.x. 0.19.3)
0.26.1 to 0.27.1

- Metabase hosting environment: (e.x. Mac app, Elastic Beanstalk, Docker, Heroku, Linux/Ubuntu 12)
docker

- Metabase internal database: (e.x. H2 (default), MySQL, Postgres)
Postgres

- *Repeatable steps to reproduce the issue*

Use docker-compose to upgrade metabase from 0.26.1 to 0.27.1, then this error occurred in the migration steps.

```
metabase22_1  | 12-04 09:37:31 INFO db.migrations :: Running all necessary data migrations, this may take a minute.
metabase22_1  | 12-04 09:37:31 INFO db.migrations :: Running data migration 'populate-card-database-id'...
metabase22_1  | org.postgresql.util.PSQLException: ERROR: insert or update on table "report_card" violates foreign key constraint "fk_report_card_ref_database_id"
metabase22_1  |   Detail: Key (database_id)=(2) is not present in table "metabase_database".
metabase22_1  | 	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2477)
metabase22_1  | 	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2190)
metabase22_1  | 	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:300)
metabase22_1  | 	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:428)
metabase22_1  | 	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:354)
metabase22_1  | 	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:169)
metabase22_1  | 	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:136)
metabase22_1  | 	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeUpdate(NewProxyPreparedStatement.java:105)
metabase22_1  | 	at clojure.java.jdbc$db_do_execute_prepared_statement$fn__10886.invoke(jdbc.clj:901)
metabase22_1  | 	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:714)
metabase22_1  | 	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
metabase22_1  | 	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:697)
metabase22_1  | 	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
metabase22_1  | 	at clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:900)
metabase22_1  | 	at clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:894)
metabase22_1  | 	at clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:932)
metabase22_1  | 	at clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:912)
metabase22_1  | 	at clojure.java.jdbc$execute_BANG_$execute_helper__10945.invoke(jdbc.clj:1161)
metabase22_1  | 	at clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1165)
metabase22_1  | 	at clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1145)
metabase22_1  | 	at toucan.db$execute_BANG_.invokeStatic(db.clj:324)
metabase22_1  | 	at toucan.db$execute_BANG_.doInvoke(db.clj:319)
metabase22_1  | 	at clojure.lang.RestFn.invoke(RestFn.java:410)
metabase22_1  | 	at toucan.db$update_BANG_.invokeStatic(db.clj:378)
metabase22_1  | 	at toucan.db$update_BANG_.invoke(db.clj:367)

▽
---
metabase22_1  | 	at toucan.db$update_where_BANG_.invokeStatic(db.clj:402)
metabase22_1  | 	at toucan.db$update_where_BANG_.doInvoke(db.clj:392)
metabase22_1  | 	at clojure.lang.RestFn.invoke(RestFn.java:464)
metabase22_1  | 	at metabase.db.migrations$populate_card_database_id.invokeStatic(migrations.clj:336)
metabase22_1  | 	at metabase.db.migrations$populate_card_database_id.invoke(migrations.clj:332)
metabase22_1  | 	at metabase.db.migrations$run_migration_if_needed_BANG_.invokeStatic(migrations.clj:49)
metabase22_1  | 	at metabase.db.migrations$run_migration_if_needed_BANG_.invoke(migrations.clj:40)
metabase22_1  | 	at metabase.db.migrations$run_all_BANG_.invokeStatic(migrations.clj:68)
metabase22_1  | 	at metabase.db.migrations$run_all_BANG_.invoke(migrations.clj:62)
metabase22_1  | 	at clojure.lang.Var.invoke(Var.java:375)
metabase22_1  | 	at metabase.db$run_data_migrations_BANG_.invokeStatic(db.clj:374)
metabase22_1  | 	at metabase.db$run_data_migrations_BANG_.invoke(db.clj:369)
metabase22_1  | 	at metabase.db$setup_db_BANG_.invokeStatic(db.clj:385)
metabase22_1  | 	at metabase.db$setup_db_BANG_.doInvoke(db.clj:376)
metabase22_1  | 	at clojure.lang.RestFn.invoke(RestFn.java:421)
metabase22_1  | 	at metabase.core$init_BANG_.invokeStatic(core.clj:108)
metabase22_1  | 	at metabase.core$init_BANG_.invoke(core.clj:87)
metabase22_1  | 	at metabase.core$start_normally.invokeStatic(core.clj:194)
metabase22_1  | 	at metabase.core$start_normally.invoke(core.clj:188)
metabase22_1  | 	at metabase.core$_main.invokeStatic(core.clj:292)
metabase22_1  | 	at metabase.core$_main.doInvoke(core.clj:287)
metabase22_1  | 	at clojure.lang.RestFn.invoke(RestFn.java:397)
metabase22_1  | 	at clojure.lang.AFn.applyToHelper(AFn.java:152)
metabase22_1  | 	at clojure.lang.RestFn.applyTo(RestFn.java:132)
metabase22_1  | 	at metabase.core.main(Unknown Source)
metabase22_1  | 12-04 09:37:31 ERROR metabase.core :: Metabase Initialization FAILED:  ERROR: insert or update on table "report_card" violates foreign key constraint "fk_report_card_ref_database_id"
metabase22_1  |   Detail: Key (database_id)=(2) is not present in table "metabase_database".
metabase22_1  | 12-04 09:37:31 INFO metabase.core :: Metabase Shutting Down ...
metabase22_1  | 12-04 09:37:31 INFO metabase.core :: Metabase Shutdown COMPLETE
```

**Comment by senior** [comment]
@sailxjx @chanthus I think I've figured out what's going wrong here.

The bug that this migration was fixing was (#5998). It's related to the database_id of native queries not being populated. Our delete logic for databases uses `database_id` to delete any related questions. I think what happened is you had native questions with no `database_id`, then deleted the associated database. Since the `database_id` of the question wasn't populated, it wasn't deleted. Unfortunately that question is broken now too, you'll get an error that says `Unable to resolve driver for query`.

I need to get a fix together for this, but it is something that you can clean up manually if you would rather not wait for the fix. The below query will show you how many questions you have in this state along with their name and whether or not they are archived:
``` sql
select id, name, archived 
from report_card 
where to_number(json_extract_path_text(dataset_query::json,'database'),'9') = <DATABASE ID>;
```
Note that `DATABASE ID` is the ID referenced in the returned error message. In the error message above, it is `2` and can be found in the detail info `Detail: Key (database_id)=(2) `.

Using a similar query you can remove the questions associated to that deleted database:
``` sql
delete from report_card 
where to_number(json_extract_path_text(dataset_query::json,'database'),'9') = <DATABASE ID>;
```

**Comment by chanthus** [comment]
Trying to delete the question results in another foreign key constraint error.

```
ERROR:  update or delete on table "report_card" violates foreign key constraint
"fk_dashboardcard_ref_card_id" on table "report_dashboardcard"
DETAIL:  Key (id)=(60) is still referenced from table "report_dashboardcard".
```

I think I'd wait for this bug to be fixed 'properly' before trying to upgrade.

**Comment by oxplot** [comment]
We had the same issue and the fix by @senior worked for us. Thanks.

**Comment by oxplot** [comment]
We had the same issue here and we just went through the questions returned by the select query:

```sql
select id, name, archived 
from report_card 
where to_number(json_extract_path_text(dataset_query::json,'database'),'9') = <DATABASE ID>;
```

and changed their database to an available one through the UI and managed to upgrade successfully.

**Comment by sunshineo** [comment]
Is this the issue I run into at https://github.com/metabase/metabase-buildpack/issues/3 ?

**Comment by senior** [comment]
@sunshineo No, but I think https://github.com/metabase/metabase/issues/6463 is the same issue you're having

**Comment by Microsheep** [comment]
Same issue here I think.

Your databases: (e.x. MySQL, Postgres, MongoDB, …)
- Google Big Query, Google Analytics, MySQL, SQLite

- Metabase version: (e.x. 0.19.3)
0.26.2 to 0.27.1

Metabase hosting environment: (e.x. Mac app, Elastic Beanstalk, Docker, Heroku, Linux/Ubuntu 12)
- docker

Metabase internal database: (e.x. H2 (default), MySQL, Postgres)
- H2

Repeatable steps to reproduce the issue
- Re-lauching docker instance after pulling newest docker version

Docker Log with Error Msg:
```
12-07 14:20:58 INFO metabase.core :: Starting Metabase in STANDALONE mode
12-07 14:20:58 INFO metabase.core :: Launching Embedded Jetty Webserver with config:
 {:port 3000, :host "0.0.0.0"}

12-07 14:20:58 INFO metabase.core :: Starting Metabase version v0.27.1 (055e448 release-0.27.0) ...
12-07 14:20:58 INFO metabase.core :: System timezone is 'Asia/Taipei' ...
12-07 14:20:59 WARN metabase.driver :: No -init-driver function found for 'metabase.driver.google'
12-07 14:20:59 INFO metabase.core :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...
12-07 14:20:59 INFO metabase.db :: Verifying h2 Database Connection ...
12-07 14:21:00 INFO metabase.db :: Verify Database Connection ...  ✅
12-07 14:21:00 INFO metabase.db :: Running Database Migrations...
12-07 14:21:00 INFO metabase.db :: Setting up Liquibase...
12-07 14:21:00 INFO metabase.db :: Liquibase is ready.
12-07 14:21:00 INFO metabase.db :: Checking if Database has unrun migrations...
12-07 14:21:01 INFO metabase.db :: Database has unrun migrations. Waiting for migration lock to be cleared...
12-07 14:21:01 INFO metabase.db :: Migration lock is cleared. Running migrations...
12-07 14:21:03 INFO metabase.db :: Database Migrations Current ...  ✅
com.mchange.v2.cfg.DelayedLogItem [ level -> FINE, text -> "The configuration file for resource identifier 'hocon:/r
eference,/application,/c3p0,/' could not be found. Skipping.", exception -> null]
12-07 14:21:04 INFO db.migrations :: Running all necessary data migrations, this may take a minute.
12-07 14:21:04 INFO db.migrations :: Running data migration 'populate-card-database-id'...
org.h2.jdbc.JdbcSQLException: Referential integrity constraint violation: "FK_REPORT_CARD_REF_DATABASE_ID: PUBLIC.RE
PORT_CARD FOREIGN KEY(DATABASE_ID) REFERENCES PUBLIC.METABASE_DATABASE(ID) (2)"; SQL statement:
UPDATE "REPORT_CARD" SET "DATABASE_ID" = 2 WHERE ("ID" in (214, 243, 222, 234, 238, 237, 240, 220, 241, 229, 228, 21
0, 211, 257, 264, 266, 267)) [23506-194]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
        at org.h2.message.DbException.get(DbException.java:179)
        at org.h2.message.DbException.get(DbException.java:155)
        at org.h2.constraint.ConstraintReferential.checkRowOwnTable(ConstraintReferential.java:371)
        at org.h2.constraint.ConstraintReferential.checkRow(ConstraintReferential.java:313)
        at org.h2.table.Table.fireConstraints(Table.java:976)
        at org.h2.table.Table.fireAfterRow(Table.java:994)
        at org.h2.command.dml.Update.update(Update.java:151)
        at org.h2.command.CommandContainer.update(CommandContainer.java:101)
        at org.h2.command.Command.executeUpdate(Command.java:258)
        at org.h2.jdbc.JdbcPreparedStatement.executeUpdateInternal(JdbcPreparedStatement.java:160)
        at org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:146)
        at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeUpdate(NewProxyPreparedStatement.java:105)
        at clojure.java.jdbc$db_do_execute_prepared_statement$fn__10886.invoke(jdbc.clj:901)
        at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:714)
        at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
        at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:697)
        at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:684)
        at clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:900)
        at clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:894)
        at clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:932)
        at clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:912)
        at clojure.java.jdbc$execute_BANG_$execute_helper__10945.invoke(jdbc.clj:1161)
        at clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1165)
        at clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1145)
        at toucan.db$execute_BANG_.invokeStatic(db.clj:324)
        at toucan.db$execute_BANG_.doInvoke(db.clj:319)
        at clojure.lang.RestFn.invoke(RestFn.java:410)
        at toucan.db$update_BANG_.invokeStatic(db.clj:378)
        at toucan.db$update_BANG_.invoke(db.clj:367)
        at toucan.db$update_where_BANG_.invokeStatic(db.clj:402)
        at toucan.db$update_where_BANG_.doInvoke(db.clj:392)
        at clojure.lang.RestFn.invoke(RestFn.java:464)
        at metabase.db.migrations$populate_card_database_id.invokeStatic(migrations.clj:336)
        at metabase.db.migrations$populate_card_database_id.invoke(migrations.clj:332)
        at metabase.db.migrations$run_migration_if_needed_BANG_.invokeStatic(migrations.clj:49)
        at metabase.db.migrations$run_migration_if_needed_BANG_.invoke(migrations.clj:40)
        at metabase.db.migrations$run_all_BANG_.invokeStatic(migrations.clj:68)
        at metabase.db.migrations$run_all_BANG_.invoke(migrations.clj:62)
        at clojure.lang.Var.invoke(Var.java:375)
        at metabase.db$run_data_migrations_BANG_.invokeStatic(db.clj:374)
        at metabase.db$run_data_migrations_BANG_.invoke(db.clj:369)
        at metabase.db$setup_db_BANG_.invokeStatic(db.clj:385)
        at metabase.db$setup_db_BANG_.doInvoke(db.clj:376)
        at clojure.lang.RestFn.invoke(RestFn.java:421)
        at metabase.core$init_BANG_.invokeStatic(core.clj:108)
        at metabase.core$init_BANG_.invoke(core.clj:87)
        at metabase.core$start_normally.invokeStatic(core.clj:194)
        at metabase.core$start_normally.invoke(core.clj:188)
        at metabase.core$_main.invokeStatic(core.clj:292)
        at metabase.core$_main.doInvoke(core.clj:287)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at metabase.core.main(Unknown Source)
12-07 14:21:04 ERROR metabase.core :: Metabase Initialization FAILED:  Referential integrity constraint violation: "
FK_REPORT_CARD_REF_DATABASE_ID: PUBLIC.REPORT_CARD FOREIGN KEY(DATABASE_ID) REFERENCES PUBLIC.METABASE_DATABASE(ID) 
(2)"; SQL statement:
UPDATE "REPORT_CARD" SET "DATABASE_ID" = 2 WHERE ("ID" in (214, 243, 222, 234, 238, 237, 240, 220, 241, 229, 228, 21
0, 211, 257, 264, 266, 267)) [23506-194]
12-07 14:21:04 INFO metabase.core :: Metabase Shutting Down ...
12-07 14:21:04 INFO metabase.core :: Metabase Shutdown COMPLETE
```

Thanks for the help!
Love metabase!

**Comment by salsakran** [comment]
@sailxjx @Microsheep @sunshineo @oxplot 

Apologies for the bug. We didn't hit it in upgrading our own instances. v0.27.2 is out and should fix this. 

Fixed via  #6568

**Comment by Microsheep** [comment]
Thanks for the help!
